src/main/java/com/google/devtools/build/lib/analysis/config/transitions/ComposingTransition.java

```diff
@@ -84,6 +79,42 @@ public boolean equals(Object other) {
         && ((ComposingTransition) other).transition2.equals(this.transition2);
   }
 
+  /**
+   * Creates a {@link ComposingTransition} that applies the sequence: {@code fromOptions ->
+   * transition1 -> transition2 -> toOptions }.
+   *
+   * <p>Note that this method checks for transitions that cannot be composed, such as if one of the
+   * transitions is {@link NoTransition} or the host transition, and returns an efficiently composed
+   * transition.
+   */
+  public static ConfigurationTransition of(
+      ConfigurationTransition transition1, ConfigurationTransition transition2) {
+    Preconditions.checkNotNull(transition1);
+    Preconditions.checkNotNull(transition2);
+
+    if (transition1 == NullTransition.INSTANCE || transition1.isHostTransition()) {
```

Comment:
Is there some way we can keep a single source of truth for isFinal()? i.e. what happens if isFinal() is changed for some TransitionFactory but this logic doesn't get updated?

Less strongly, can we at least keep the isFinal method in here rather than explicitly calling out NullTransition and HostTransition twice?

Reply:
There is currently no isFinal for ConfigurationTransition instances, so the only way to handle this is by explicit checks of specific transition types.

Maybe this is actually cleaner? I could remove isFinal from TransitionFactory and just check specific types in ComposingTransitionFactory instead. I think this is the only place where isFinal is used.

Reply:
I'm actually okay either way - keeping it in the TransitionFactory definition or centrally checking.

Whichever way though, let's avoid explicitly checking for NullTransition and HostTransition twice or more. For example, defining isFinal() in ComposingTransition and calling that form ComposingTransition.of and ComposingTransitionFactory.   

If isFinal remains in TransitionFactory I'm quite sure how to access that source of truth in ComposingTransition.

Reply:
Removed isFinal from TransitionFactory, gave ComposingTransition and ComposingTransitionFactory better isFinal methods for checking.